// @ngx-i18nsupport/ngx-i18nsupport-lib/esm5/api/i-icu-message.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * A message category, which is a part of an ICU message.
 * E.g. the ICU message {wolves, plural, =0 {no wolves} =1 {one wolf} =2 {two wolves} other {a wolf pack}}
 * has 4 category objects with the categories =0, =1, =2, other.
 * @record
 */
export function IICUMessageCategory() { }
if (false) {
    /**
     * Fix part.
     * For plural mesages the category is "=0" or "=1" or "few" or "other"...
     * For select messages it is the matching key.
     * @return {?}
     */
    IICUMessageCategory.prototype.getCategory = function () { };
    /**
     * Translatable part.
     * @return {?}
     */
    IICUMessageCategory.prototype.getMessageNormalized = function () { };
}
/**
 * An ICU message.
 * @record
 */
export function IICUMessage() { }
if (false) {
    /**
     * Is it a plural message?
     * @return {?}
     */
    IICUMessage.prototype.isPluralMessage = function () { };
    /**
     * Is it a select message?
     * @return {?}
     */
    IICUMessage.prototype.isSelectMessage = function () { };
    /**
     * All the parts of the message.
     * E.g. the ICU message {wolves, plural, =0 {no wolves} =1 {one wolf} =2 {two wolves} other {a wolf pack}}
     * has 4 category objects with the categories =0, =1, =2, other.
     * @return {?}
     */
    IICUMessage.prototype.getCategories = function () { };
    /**
     * Returns the icu message content as format dependent native string.
     * This is, how it is stored, something like '{x, plural, =0 {..}'
     * @return {?}
     */
    IICUMessage.prototype.asNativeString = function () { };
    /**
     * Translate message and return a new, translated message
     * @param {?} translation the translation (hashmap of categories and translations).
     * @return {?} new message wit translated content.
     */
    IICUMessage.prototype.translate = function (translation) { };
}
