// @ngx-i18nsupport/ngx-i18nsupport-lib/esm2015/api/i-normalized-message.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Created by martin on 09.05.2017.
 * A normalized message is an abstraction of a translation containing some markup.
 * Markup can be placeholders or html tags.
 * @record
 */
export function ValidationErrors() { }
if (false) {
    /** @type {?|undefined} */
    ValidationErrors.prototype.placeholderAdded;
    /** @type {?|undefined} */
    ValidationErrors.prototype.placeholderRemoved;
    /** @type {?|undefined} */
    ValidationErrors.prototype.tagAdded;
    /** @type {?|undefined} */
    ValidationErrors.prototype.tagRemoved;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function INormalizedMessage() { }
if (false) {
    /**
     * normalized message as string.
     * @param {?=} format optional way to determine the exact syntax.
     * Allowed formats are defined as constants NORMALIZATION_FORMAT...
     * @return {?}
     */
    INormalizedMessage.prototype.asDisplayString = function (format) { };
    /**
     * Validate the message.
     * @return {?} null, if ok, error object otherwise.
     */
    INormalizedMessage.prototype.validate = function () { };
    /**
     * Validate the message, check for warnings only.
     * A warning shows, that the message is acceptable, but misses something.
     * E.g. if you remove a placeholder or a special tag from the original message, this generates a warning.
     * @return {?} null, if no warning, warnings as error object otherwise.
     */
    INormalizedMessage.prototype.validateWarnings = function () { };
    /**
     * Returns the message content as format dependent native string.
     * Includes all format specific markup like <ph id="INTERPOLATION" ../> ..
     * @return {?}
     */
    INormalizedMessage.prototype.asNativeString = function () { };
    /**
     * Test wether this message is an ICU message.
     * @return {?} true, if it is an ICU message.
     */
    INormalizedMessage.prototype.isICUMessage = function () { };
    /**
     * Test wether this message contains an ICU message reference.
     * ICU message references are something like <x ID="ICU"../>.
     * @return {?} true, if there is an ICU message reference in the message.
     */
    INormalizedMessage.prototype.containsICUMessageRef = function () { };
    /**
     * If this message is an ICU message, returns its structure.
     * Otherwise this method returns null.
     * @return {?} ICUMessage or null.
     */
    INormalizedMessage.prototype.getICUMessage = function () { };
    /**
     * Create a new normalized message as a translation of this one.
     * @throws an error if normalized string is not well formed.
     * Throws an error too, if this is an ICU message.
     * @param {?} normalizedString the translation in normalized form.
     * If the message is an ICUMessage (getICUMessage returns a value), use translateICUMessage instead.
     * @return {?}
     */
    INormalizedMessage.prototype.translate = function (normalizedString) { };
    /**
     * Create a new normalized icu message as a translation of this one.
     * @throws an error if normalized string is not well formed.
     * Throws an error too, if this is not an ICU message.
     * @param {?} icuTranslation the translation, this is the translation of the ICU message,
     * which is not a string, but a collections of the translations of the different categories.
     * The message must be an ICUMessage (getICUMessage returns a value)
     * @return {?}
     */
    INormalizedMessage.prototype.translateICUMessage = function (icuTranslation) { };
    /**
     * Create a new normalized message from a native xml string as a translation of this one.
     * @param {?} nativeString xml string in the format of the underlying file format.
     * Throws an error if native string is not acceptable.
     * @return {?}
     */
    INormalizedMessage.prototype.translateNativeString = function (nativeString) { };
}
